### ISSUE TITLE
Allow hosts to access variables from other nodes

### ISSUE THREAD

**Issue by mpdehaan** [ansible/ansible#130]
1. Implement the fetch module which stores facts per host in a specific directory
2. After setup runs, the host variables from each host are stored on the filesystem in a known directory structure, probably ~/.ansible_setup_variables/host.json or similar
3.   Add a custom Jinja2 extension is made that allows access to something like getvar(hostname, key)  and make sure our template module uses it
4.  We shift the template module to happen entirely server side (for evaluation), eliminating the need to execute templates on the host, and on the host (via Runner code) we only need to use the COPY module (not template) to move the file.   Now we may have to continue to support the local template evaluation for regular Ansible, since we've documented that you're allowed to call setup explicitly, so this is probably something only for playbooks, and we make the equivalent of template_local to do this.

RESULT -- we can do cool things like allow the webservers to know the ip address of the monitoring server and auto configure, etc.

Should be pretty simple following the above steps.

**Comment by mpdehaan** [comment]
I've implemented 2 without needing to rely on the fetch module, so it's faster.

Files are saved by default in /var/lib/ansible/setup_data/hostname/setup or if run as non-root ~./ansible/setup_data/hostname/setup

We will be using this file with the template module in 4.   Runner will respond to "is_playbook=True" by running local templates versus remote templates, but this won't occur for non-playbooks.  When it uses local templates, it will create temporary files
and use the 'copy' module instead of the template module, then delete the temp file once transferred

Steps 3 and 4 are now unblocked and ready to proceed, and 4 has advantages (less-deps) even w/o 3.

**Comment by mpdehaan** [comment]
#152 is opened as a prereq for the local templates

**Comment by dagwieers** [comment]
Is there a caching mechanism in order to not recreate "dynamic" facts every time ? Some facts may not change without a reboot, or do not change very often, or will never change. In order to make execution faster we may want to cache facts for a certain time ? A caching mechanism would be useful both for use on the host as on nodes.

**Comment by mpdehaan** [comment]
Dag -- I don't understand, are you talking about facter/ohai?  In which case they do their own thing... or the proposal to add our own?   Seems to not be about this ticket, apologies for confusion.

This is about storing the variables from the host when setup runs, so it will only happen once per node and then make those variables available to templates, so I think, basically, it does what you want... facts get evaluated once per playbook.  If you are using vars_files, actually twice.  But not three times.  Three is right out.

**Comment by dagwieers** [comment]
I am talking about our own facts implementation. Since we have no daemon to take care of this, I would expect ansible to trigger in certain cases an update of facts. And this is related to this issue, because if you store facts on the host, you need to know when these "cached" values need to be refreshed.

I am assuming that you like to cache facts on hosts too, rather than fetch (and regenerate if not cached on the node) them for all systems all the time.

PS My preference is to not rely on facter/ohai at all, so I am looking mostly to that scenario.

**Comment by mpdehaan** [comment]
yeah, triggering/update will happen once per play.   That's the extent of the caching and should be more than good enough.   They are not accessed on demand.

**Comment by dagwieers** [comment]
Ok, in case these are Ansible facts, are they being "executed" at the time of need as well ? Or are we only considering static facts ? (I guess this discussion then should be part of another issue ?)

**Comment by mpdehaan** [comment]
the setup module will contain a lot of python code to automatically return some extra variables, all the time, more or less.  There will not be additional external facts in the initial implementation.   Once we change the way content is transferred (to the tarball approach in another ticket, probably) that can be revisited to support executable facts.

**Comment by mpdehaan** [comment]
all templates are now local, remaining step is just the jinja2 extension to access the other host's variables.  This should only be legal for playbooks and probably needs to leverage SETUP_CACHE for now

**Comment by mpdehaan** [comment]
this is now done on the integration branch, see email to list.
